Convention-based handlers (#8224)
Particular/docs.particular.net: nservicebus/handlers/convention-based.md

---
title: Convention-based handlers
summary: How to create message handlers that don't implement IHandleMessages<T>
component: Core
versions: '[10,)'
reviewed: 2026-05-04
related:
---

Starting with NServiceBus version 10.2.0, NServiceBus supports **convention-based message handlers** that do not implement `IHandleMessages<T>`. These handlers are not discovered through traditional assembly scanning, but instead are either added to an NServiceBus endpoint declaratively, or with help from Roslyn analyzers and source generators.

## Handler structure

> A convention-based handler can look exactly like a regula rmessage handler, but does not implement the interface:

Reviewer: ```suggestion
A convention-based handler can look exactly like a regular message handler, but does not implement the interface:
```

snippet: SimpleConventionBasedHandler

Without the rigid structure of the `IHandleMessages<T>` interface, additional parameters can be added to the `Handle` method:

snippet: ConventionsBasedHandlerExtraParams

> These requirements must be met for a class to be recongizable as a convention-based message handler:

Reviewer: ```suggestion
These requirements must be met for a class to be recognized as a convention-based message handler:
```

- Must contain a handler method named `Handle` which returns `Task`.
- The handler method's first parameter must be a message class.

> - The hadnler method's second parameter must be an `IMessageHandlerContext`.

Reviewer 2: ```suggestion
- The handler method's second parameter must be an `IMessageHandlerContext`.
```

> - After the first two parameters, any additional parameters must be either a `CancellationToken` or registered in host's `IServiceCollection`.

Reviewer 2: ```suggestion
- After the first two parameters, any additional parameters must be either a `CancellationToken` or Services registered in the host's `IServiceCollection`.
```
Alternativ: dependencies

> - The method can be an instance or static method.

Reviewer 2: ```suggestion
- It can be an instance or static method.
```

To avoid repeating method?

> - A handler class may contain multiple handler methods, differing by the message type, but these methods become an inseparable unit. It is not possible to register one handler method on a class but not the other.

Reviewer 2: Would inline code demonstrating the convention be helpful to make the text more memorable?

Author: Did the new TLDR content at the top address this? If not, I'm not sure exactly what you mean.

Reviewer 2: yes

> TODO: What happens if there are 2 handler methods for the same message type but with different additional params?

Reviewer 2: Two handlers for the same message. One method = one handler. The method plus the dependencies makes up the identity as far as I recall

snippet: ConventionHandlerRegistrationWithoutAttribute

However, decorating a handler class with the `NServiceBus.HandlerAttribute` (or `NServiceBus.SagaAttribute` for [sagas](/nservicebus/sagas/)) enables source generation that enables all decorated handlers and/or sagas in an entire project to be added with one line of configuration.

First, decorate handlers with `[Handler]`, or sagas with `[Saga]`:

snippet: DecoratedConventionHandler

This generates source code that allows all handlers and sagas from an assembly to be registered at once:

snippet: ConventionHandlerAddAllFromAssembly

However, the generated source is flexible and allows registering just all handlers, just all sagas, or both from any level of the namespace hierarchy, or at the top level of the assembly:

snippet: ConventionHandlerAllGeneratedAddMethods

## Analyzers

While the source generation simplifies registering multiple handlers or sagas to an endpoint with one line of code, the generation relies on the `[Handler]` and `[Saga]` attributes to identify what qualifies as a handler or a saga.

> [!TIP]

> > The reason the source generation works on the `[Handler]` and `[Saga]` attributes is because source generators, which may run in your editor up to every time you press a key, must be very fast and efficient. Identifying generation targets using a marker attribute is the most optimized method availble when using the Roslyn SDK. On the other hand, using marker interfaces to identify generation targets is [explicitly called out as an anti-pattern in the source generator documentation](https://github.com/dotnet/roslyn/blob/main/docs/features/incremental-generators.cookbook.md).

Reviewer 2: ```suggestion
> The reason the source generation works on the `[Handler]` and `[Saga]` attributes is because source generators, which may run in your editor up to every time you press a key, must be very fast and efficient. Identifying generation targets using a marker attribute is the most optimized method available when using the Roslyn SDK. On the other hand, using marker interfaces to identify generation targets is [explicitly called out as an anti-pattern in the source generator documentation](https://github.com/dotnet/roslyn/blob/main/docs/features/incremental-generators.cookbook.md).
```

## Handler structure

A convention-based handler can look exactly like a regular message handler, but does not implement the interface:

snippet: SimpleConventionBasedHandler

Without the rigid structure of the `IHandleMessages<T>` interface, additional parameters can be added to the `Handle` method:

snippet: ConventionsBasedHandlerExtraParams

These requirements must be met for a class to be recognized as a convention-based message handler:

- Must contain a handler method named `Handle` which returns `Task`.
- The handler method's first parameter must be a message class.
- The handler method's second parameter must be an `IMessageHandlerContext`.
- After the first two parameters, any additional parameters must be either a `CancellationToken` or Services registered in the host's `IServiceCollection`.
- It can be an instance or static method.
- A handler class may contain multiple handler methods, differing by the message type, but these methods become an inseparable unit. It is not possible to register one handler method on a class but not the other.

> - If multiple handler methods use the same message type as the first parameter, they will all be executed on the same message.

Author: So is that accurate then?

Reviewer 2: yes
